### ISSUE TITLE
Using `buildOcaml` in `nix-shell` results in error about mkdir

### ISSUE THREAD

**Issue by CMCDragonkai** [NixOS/nixpkgs#51658]
## Issue description

I tried using `buildOcaml` from `ocaml-ng.ocamlPackages_4_07` and when running `nix-shell` it gave me an error about:

```
mkdir: cannot create directory '/nix/store/7mmv0hk8r13vlm8vp88qzl0dkwpppzm8-ocaml-OCaml-Demo-0.0.1': Read-only file system
```

### Steps to reproduce

```
pkgs.ocaml-ng.ocamlPackages_4_07.buildOcaml {
  name = "OCaml-Demo";
  version = "0.0.1";
  src = null;
}
```

It appears due to the way `ocamlPackages.findlib` is trying to create some find lib destination directory when `createFindlibDestdir = true;`.

**Comment by vbgl** [comment]
Out of curiosity: why did you try to use `buildOcaml`?

I guess you can explicitly set `createFindlibDestdir` to `false`.

**Comment by CMCDragonkai** [comment]
In other language envs, they all use some variant of `stdenv.mkDerivation`. But ocaml has buildOcaml so I thought it would be used.

**Comment by sternenseemann** [comment]
This should be fixed by now, see #121991.
